src/librustc_privacy/lib.rs

```diff
@@ -1321,7 +1321,18 @@ impl<'a, 'tcx, 'v> Visitor<'v> for VisiblePrivateTypesVisitor<'a, 'tcx> {
                     match *trait_ref {
                         None => {
                             for impl_item in impl_items {
-                                visit::walk_impl_item(self, impl_item);
+                                match impl_item.node {
+                                    ast::MethodImplItem(..)
+                                        if self.exported_items.contains(&impl_item.id) ||
+                                           impl_item.vis == ast::Public =>
```

Comment:
maybe we need an accessor for this compound condition (exported_items.contains || vis == pub), it appears also below...?